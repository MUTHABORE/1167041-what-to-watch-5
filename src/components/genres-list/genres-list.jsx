import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {changeActiveGenre, changeAmountMoviesToRender} from '../../store/action.js';
import {getFilteredMovies} from '../../store/selectors.js';
import {propsForFilms} from '../../util/props-validation.js';
import PropTypes from 'prop-types';

import MoviesList from '../movies-list/movies-list.jsx';
import {AMOUNT_MOVIES_TO_RENDER} from '../../util/const.js';

const GenresList = (props) => {
  const {activeGenre, moviesList, genresList, amountMoviesToRender, changeActiveGenreAction, changeAmountMoviesToRenderAction} = props;
  return (
    <>
      <ul className="catalog__genres-list">
        {genresList.map((genre, i) => (
          <li key={i} className={`catalog__genres-item ${activeGenre === genre ? `catalog__genres-item--active` : ``}`} onClick={(evt) => {
            evt.preventDefault();
            changeActiveGenreAction(genre);
            changeAmountMoviesToRenderAction(AMOUNT_MOVIES_TO_RENDER);
          }
          }>
            <Link to="#" className="catalog__genres-link">{genre}</Link>
          </li>
        ))}
      </ul>
      <MoviesList moviesList={moviesList} amountMoviesToRender={amountMoviesToRender} changeAmountMoviesToRender={changeAmountMoviesToRenderAction} />
    </>
  );
};

GenresList.propTypes = {
  moviesList: PropTypes.arrayOf(propsForFilms).isRequired,
  genresList: PropTypes.array.isRequired,
  activeGenre: PropTypes.string.isRequired,
  amountMoviesToRender: PropTypes.number.isRequired,
  changeActiveGenreAction: PropTypes.func.isRequired,
  changeAmountMoviesToRenderAction: PropTypes.func.isRequired,
};

const mapStateToProps = ({DATA, FUNCTIONAL}) => ({
  activeGenre: FUNCTIONAL.activeGenre,
  moviesList: getFilteredMovies({genresList: FUNCTIONAL.activeGenre, moviesList: DATA.moviesList}),
  genresList: DATA.genresList,
  amountMoviesToRender: FUNCTIONAL.amountMoviesToRender,
});

const mapDispatchToProps = (dispatch) => ({
  changeActiveGenreAction(genre) {
    dispatch(changeActiveGenre(genre));
  },
  changeAmountMoviesToRenderAction(amount) {
    dispatch(changeAmountMoviesToRender(amount));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
