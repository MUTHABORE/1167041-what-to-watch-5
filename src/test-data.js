import React from 'react';
import PropTypes from 'prop-types';

export const noop = () => {};

export const routerProps = {
  match: {
    params: {
      id: `1`,
    }
  }
};

export const movies = [
  {
    name: `Bronson`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    rating: 3.6,
    votes: 109661,
    director: `Nicolas Winding Refn`,
    starring: [
      `Tom Hardy`,
      `Kelly Adams`,
      `Luing Andrews`
    ],
    runtime: 92,
    genre: `Action`,
    released: 2008,
    id: 1,
    isFavorite: false,
    video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  },
  {
    name: `Bronson`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    rating: 3.6,
    votes: 109661,
    director: `Nicolas Winding Refn`,
    starring: [
      `Tom Hardy`,
      `Kelly Adams`,
      `Luing Andrews`
    ],
    runtime: 92,
    genre: `Action`,
    released: 2008,
    id: 2,
    isFavorite: true,
    video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }
];

export const reviews = [
  {
    id: 1,
    user: {
      id: 11,
      name: `Jack`
    },
    rating: 8.6,
    comment: `A movie that will take you to another world full of emotions.`,
    date: `2020-11-03T13:38:44.769Z`
  },
  {
    id: 2,
    user: {
      id: 12,
      name: `Jack`
    },
    rating: 8.6,
    comment: `A movie that will take you to another world full of emotions.`,
    date: `2020-11-03T13:38:44.769Z`
  }
];

export const movie = movies[0];

export const userMock = {
  id: 1,
  name: `GERASIM`,
  email: `moo@moo.com`,
  avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/5.jpg`,
};

export const state = {
  DATA: {
    moviesList: [movies[0], movies[1]],
    genresList: [`All genres`, `Action`],
    reviews: [],
    promo: movies[0],
  },
  FUNCTIONAL: {
    activeGenre: `All genres`,
    amountMoviesToRender: 8,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    user: userMock,
  },
};

export const MockComponent = (props) => {
  const {children} = props;
  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

export const SimpleMockComponent = () => <div />;
