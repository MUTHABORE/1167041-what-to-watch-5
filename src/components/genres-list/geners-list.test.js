import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';
import {state, noop} from '../../test-data';
import {GenresList} from './genres-list.jsx';

test(`Should GenresList render correctly`, () => {
  const tree = renderer
    .create(
        <BrowserRouter>
          <GenresList
            activeGenre={state.FUNCTIONAL.activeGenre}
            moviesList={state.DATA.moviesList}
            genresList={state.DATA.genresList}
            amountMoviesToRender={state.FUNCTIONAL.amountMoviesToRender}
            changeActiveGenreAction={noop}
            changeAmountMoviesToRenderAction={noop}
          />
        </BrowserRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
