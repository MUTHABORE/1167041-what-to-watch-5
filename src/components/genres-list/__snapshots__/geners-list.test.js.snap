// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should GenresList render correctly 1`] = `
Array [
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
      onClick={[Function]}
    >
      <a
        className="catalog__genres-link"
        href="/"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
      onClick={[Function]}
    >
      <a
        className="catalog__genres-link"
        href="/"
        onClick={[Function]}
      >
        Action
      </a>
    </li>
  </ul>,
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="/films/1"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height={175}
            loop={true}
            muted={true}
            poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg"
            preload="true"
            src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <span>
            Bronson
          </span>
        </h3>
      </a>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="/films/2"
        onClick={[Function]}
      >
        <div
          className="small-movie-card__image"
        >
          <video
            height={175}
            loop={true}
            muted={true}
            poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg"
            preload="true"
            src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
            width={280}
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <span>
            Bronson
          </span>
        </h3>
      </a>
    </article>
  </div>,
  <div
    className="catalog__more"
  >
    
  </div>,
]
`;
